src/ott/neural/solvers/flows.py

```diff
@@ -0,0 +1,250 @@
+# Copyright OTT-JAX
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import abc
+
+import jax
+import jax.numpy as jnp
+
+__all__ = [
+    "BaseFlow", "StraightFlow", "ConstantNoiseFlow", "BrownianNoiseFlow",
+    "BaseTimeSampler", "UniformSampler", "OffsetUniformSampler"
+]
+
+
+class BaseFlow(abc.ABC):
+  """Base class for all flows.
+
+  Args:
+    sigma: Constant noise used for computing time-dependent noise schedule.
+  """
+
+  def __init__(self, sigma: float) -> None:
+    self.sigma = sigma
+
+  @abc.abstractmethod
+  def compute_mu_t(self, t: jnp.ndarray, x_0: jnp.ndarray, x_1: jnp.ndarray):
+    """Compute the mean of the probablitiy path.
+
+    Compute the mean of the probablitiy path between :math:`x` and :math:`y`
+    at time :math:`t`.
+
+    Args:
+      t: Time :math:`t`.
+      x_0: Sample from the source distribution.
+      x_1: Sample from the target distribution.
+    """
+    pass
+
+  @abc.abstractmethod
+  def compute_sigma_t(self, t: jnp.ndarray):
+    """Compute the standard deviation of the probablity path at time :math:`t`.
+
+    Args:
+      t: Time :math:`t`.
+    """
+    pass
+
+  @abc.abstractmethod
+  def compute_ut(
+      self, t: jnp.ndarray, x_0: jnp.ndarray, x_1: jnp.ndarray
+  ) -> jnp.ndarray:
+    """Evaluate the conditional vector field.
+
+      Evaluate the conditional vector field defined between :math:`x_0` and
+      :math:`x_1` at time :math:`t`.
+
+    Args:
+      t: Time :math:`t`.
+      x_0: Sample from the source distribution.
+      x_1: Sample from the target distribution.
+    """
+    pass
+
+  def compute_xt(
+      self, noise: jnp.ndarray, t: jnp.ndarray, x_0: jnp.ndarray,
+      x_1: jnp.ndarray
+  ) -> jnp.ndarray:
+    """Sample from the probability path.
+
+      Sample from the probability path between :math:`x_0` and :math:`x_1` at
+      time :math:`t`.
+
+    Args:
+      noise: Noise sampled from a standard normal distribution.
```

Comment:
Because we need to "optimally" match the noise with the data, hence the noise is ordered here. We could obviously match the noise with the data within this step, but IMO this doesn't fit nicely here.